File: DemoServer/AdditionalLanguages/python/facetedSearch/facetsOptions/facetsOptions.py
from demo_example import Example, RunParamsBase
from models import Product

#region Usings
from ravendb import AbstractIndexCreationTask, Facet
from ravendb.documents.queries.facets.misc import FacetOptions, FacetTermSortMode
#endregion

class RunParams(RunParamsBase):
    def __init__(
        self,
        start: int = None,
        page_size: int = None,
        include_remaining_terms: str = None,
    ):
        self.start = start
        self.page_size = page_size
        self.include_remaining_terms = include_remaining_terms


#region Demo
#region Step_1
class Products_ByCategoryAndSupplier(AbstractIndexCreationTask):
    def __init__(self):
        super().__init__()
        self.map = (
            "docs.Products.Select(product => new {"
            '    CategoryName = (this.LoadDocument(product.Category, "Categories")).Name,'
            "    Supplier = product.Supplier"
            "})"
        )
#endregion
#endregion

class FacetsOptions(Example):
    def run(self, run_params: RunParams):
        start = run_params.start or 3
        pageSize = run_params.page_size or 2
        include = run_params.include_remaining_terms
        includeRemainingTerms = include is None or include == "true"
        
        Products_ByCategoryAndSupplier().execute(self.document_store_holder.store())
        #region Demo
        #region Step_2
        facet1 = Facet()
        facet1.field_name = "CategoryName"

        facet1_options = FacetOptions()
        facet1_options.start = start
        facet1_options.page_size = pageSize
        facet1_options.include_remaining_terms = includeRemainingTerms
        facet1_options.term_sort_mode = FacetTermSortMode.COUNT_DESC
        
        facet1.options = facet1_options
        #endregion

        facet2 = Facet()
        facet2.field_name = "Supplier"

        facet2_options = FacetOptions()
        facet2_options.start = start
        facet2_options.page_size = pageSize
        facet2_options.include_remaining_terms = includeRemainingTerms
        facet2_options.term_sort_mode = FacetTermSortMode.VALUE_ASC

        facet2.options = facet2_options
        
        #region Step_3
        facets = [facet1, facet2]
        #endregion

        with self.document_store_holder.store().open_session() as session:
            #region Step_4
            query_results = (
                session.query_index_type(Products_ByCategoryAndSupplier, Product)
                    .aggregate_by_facets(*facets)
                    .execute()
            )
            #endregion
        #endregion

        return query_results

File: DemoServer/AdditionalLanguages/php/facetedSearch/facetsOptions/FacetsOptions.php
<?php

namespace RavenDB\Demo\facetedSearch\facetsOptions;

//region Usings
use RavenDB\Documents\Indexes\AbstractIndexCreationTask;
use RavenDB\Documents\Queries\Facets\Facet;
use RavenDB\Documents\Queries\Facets\FacetOptions;
use RavenDB\Documents\Queries\Facets\FacetResultArray;
use RavenDB\Documents\Queries\Facets\FacetTermSortMode;
//endregion

use RavenDB\Demo\common\DocumentStoreHolder;
use RavenDB\Demo\common\models\Product;

//region Demo
//region Step_1
class Products_ByCategoryAndSupplier extends AbstractIndexCreationTask
{
    public function __construct()
    {
        parent::__construct();

        $this->map = "docs.Products.Select(product => new {" .
            "    CategoryName = (this.LoadDocument(product.Category, \"Categories\")).Name," .
            "    Supplier = product.Supplier" .
            "})";
    }
}
//endregion
//endregion

class FacetsOptions
{
    public function __invoke(RunParams $runParams): FacetResultArray
    {
        $start = $runParams->getStart() ?? 3;
        $pageSize = $runParams->getPageSize() ?? 2;

        $include = $runParams->getIncludeRemainingTerms();
        $includeRemainingTerms = $include == null || "true" == $include;

        //region Demo
        //region Step_2
        $facet1 = new Facet();
        $facet1->setFieldName("CategoryName");

        $facet1Options = new FacetOptions();
        $facet1Options->setStart($start);
        $facet1Options->setPageSize($pageSize);
        $facet1Options->setIncludeRemainingTerms($includeRemainingTerms);
        $facet1Options->setTermSortMode(FacetTermSortMode::countDesc());
        $facet1->setOptions($facet1Options);
        //endregion

        $facet2 = new Facet();
        $facet2->setFieldName("Supplier");

        $facet2Options = new FacetOptions();
        $facet2Options->setStart($start);
        $facet2Options->setPageSize($pageSize);
        $facet2Options->setIncludeRemainingTerms($includeRemainingTerms);
        $facet2Options->setTermSortMode(FacetTermSortMode::valueAsc());
        $facet2->setOptions($facet2Options);

        //region Step_3
        $facets = [ $facet1, $facet2 ];
        //endregion

        $queryResults = [];

        $session = DocumentStoreHolder::getStore()->openSession();
        try {
            //region Step_4
            $queryResults = $session->query(Product::class, Products_ByCategoryAndSupplier::class)
                ->aggregateBy($facets)
                ->execute();
            //endregion
        } finally {
            $session->close();
        }
        //endregion

        return $queryResults;
    }
}
